### ISSUE TITLE
importccl: IMPORT INTO should support resolving default expressions referencing UDTs

### ISSUE THREAD

**Issue by ajstorm** [cockroachdb/cockroach#61133]
Currently IMPORT doesn't work for tables with computed columns that use user defined types.  For example, this test:

```
CREATE TYPE typ as ENUM ('open', 'closed'); 
CREATE TABLE typ_table (i INT8 PRIMARY KEY, t typ DEFAULT 'open', s text, b bytea)`)
IMPORT INTO typ_table (i, s, b) AVRO DATA nodelocal://0/simple.ocf`
```

fails with:

error executing 'IMPORT INTO typ_table (i, s, b) AVRO DATA nodelocal://0/simple.ocf': pq: nodelocal://0/simple.ocf: process default and computed columns: type OID 100058 does not exist

This is also needed to allow for IMPORTing into the new multi-region REGIONAL BY ROW tables, as they use a hidden crdb_region column under the covers, which is based on the crdb_internal_region system defined type.

Epic CRDB-7074

**Comment by adityamaru** [comment]
Today, `IMPORT INTO` does not support default expressions with UDTs and will fail when attempting to resolve the UDT because of a nil type resolver in the `semaCtx`. To support UDT default expressions we need to do a couple of things:

1. Resolve all UDT default expressions during import planning, and save the type descriptors in the import job. This information will capture all the types that will be read during the execution of the import job.

2. When producing datums for the `IMPORT INTO` we will synthesize our own resolver that is populated using information from these cached type descriptors. This means that import job execution will always have a consistent view of the types that were read when the import was planned.

3. At the end of the IMPORT, prior to making the tables public, we will iterate over our cached descriptors and ensure that their version number has not changed during the execution of the import job. This guarantees that the type has not changed from under us (shrunk/grown) during the import and importing keys is safe. If the version numbers are mismatched, we fail the import into and rollback all our ingested keys.

**Comment by adityamaru** [comment]
This is fixed and backported to 21.2!
